transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -156,14 +180,41 @@ public void run() {
         return promise;
     }
 
-    private static void notifyConnect(ChannelFuture future, Promise<Channel> promise) {
+    private void notifyConnect(Future<Channel> future, final Promise<Channel> promise) {
         if (future.isSuccess()) {
-            promise.setSuccess(future.channel());
+            final Channel channel = future.getNow();
+            assert !channel.config().isAutoRead();
+            channel.attr(POOL_KEY).set(this);
+
+            if (channel.eventLoop().inEventLoop()) {
+                notifyChannelCreated(channel, promise);
+            } else {
+                // Must be called in the EventLoop of the Channel so submit to the executor.
+                channel.eventLoop().execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        notifyChannelCreated(channel, promise);
+                    }
+                });
+            }
         } else {
             promise.setFailure(future.cause());
         }
     }
 
+    private void notifyChannelCreated(Channel channel, Promise<Channel> promise) {
+        assert channel.eventLoop().inEventLoop();
+        try {
+            // Mark as autoRead as after the channelCreated(...) method is called the handlers should have been set up.
+            channel.config().setAutoRead(true);
```

Comment:
no... the problem is that how handler.channelCreated(...) is now called. Before it was in channelRegistered(...) and so the user was able to ensure all handlers are added before something is read from the channel. Now its after connect is successful so it may be too late. I'm not 100% sure yet if there is a better way. Let me think about it